File: IPWorks Samples/IMAP Email Client/javascript/imap.js
/*
 * IPWorks 2022 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const ipworks = require("@nsoftware/ipworks");

if(!ipworks) {
  console.error("Cannot find ipworks.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

main();

async function main() {
	const argv = process.argv;
	if(argv.length < 8 || argv.length > 11) {
		console.log("Usage: node imap.js -s server -u username -p password [-option parameter]]");
		console.log("Options: ");
		console.log("  -ssl     a switch parameter to enable SSL");
		console.log("  -s       the name or address of the mail server (IMAP server)");
		console.log("  -port    the port of the mail server");
		console.log("  -u       the username used to authenticate to the mail server");			
		console.log("  -p       the password used to authenticate to the mail server ");
		console.log("Example: node imap.js -s my.imapserver.com -port 993 -u username -p password -ssl");
		process.exit();
	}

	let count = 0;
	let mailboxes=[];
	let messageCount=0;
	const imap = new ipworks.imap();
  let server, port, username, password;

	for (i = 0; i < argv.length; i++) {
		if (argv[i].startsWith("-")) {
			if (argv[i] === "-ssl") {  imap.setSSLEnabled(true); } 
			if (argv[i] === "-s") {    server = argv[i + 1]; }
			if (argv[i] === "-port") { port = argv[i + 1]; }
			if (argv[i] === "-u") {    username = argv[i + 1]; } 
			if (argv[i] === "-p") {    password = argv[i + 1]; }
		}
	}

	imap.on("MailboxList", function(e){
		count+=1;
		mailboxes.push(e.mailbox);
	})
	.on("MessageInfo", function(e){
		console.log(e.messageId + ") " + e.subject);
	})
	.on("Transfer", function(e){
		console.log(e.text.toString());
	})
	.on("SSLServerAuthentication", function(e) {
		e.accept = true;
	});

	function displayMailBoxList(){
		for (i=1; i<count+1; i++){
			console.log(i + ")" + mailboxes[i-1]);
		}   
	}

	try {
		imap.setMailServer(server);
		if (port) { imap.setMailPort(parseInt(port)); }
		imap.setUser(username);
		imap.setPassword(password);
		await imap.connect();
		imap.setMailbox("*");
		await imap.listMailboxes();
	} catch (ex) {
		console.log("Error: "+ex);
		return;
	}

	console.log('Welcome! Type ? for a list of commands.');
	process.stdout.write('imap> ');



	rl.on('line', async function(line) {
		var index=line.toLowerCase().split(/\s+/)[1];
		if(line.toLowerCase() === '?') {
			console.log("IMAP Demo Commands");
			console.log("l                 list mailboxes");
			console.log("s mailbox_num     show messages in mailbox by index eg:s 1");
			console.log("r message_num     read the message by index eg:r 1");
			console.log("q                 quit");
		} else if (line.toLowerCase().startsWith('l')){
			console.log("Mailboxes for " + argv[2] + ":");
			count = 0;
			mailboxes=[];
			imap.setMailbox("*");
			try {
				await imap.listMailboxes();
				displayMailBoxList();
				console.log("Total: " + count.toString());
			} catch (ex) {
				console.log("Error: "+ex);
			}
		} else if (line.toLowerCase().startsWith('s')){		
			imap.setMailbox(mailboxes[parseInt(index)-1]);
			try {
				await imap.selectMailbox();
				messageCount=parseInt(imap.getMessageCount());
				if (messageCount!==0) {
					imap.setMessageSet("1:" + messageCount.toString());
					console.log("Messages in current mailbox:");
					await imap.fetchMessageInfo();
				} else{
					console.log("No messages found in the current mailbox.");
				}
			} catch(ex) {
				console.log("Error: "+ex);
			}
		} else if (line.toLowerCase().startsWith('r')){
			var currentMailbox=imap.getMailbox();
			messageCount=parseInt(imap.getMessageCount());
			if (currentMailbox !='*') {		
				if (parseInt(index)<=messageCount) {
					console.log("Message text:");
					imap.setMessageSet(index);
					await imap.fetchMessageText().catch((err) => console.log("Error: "+err.message));;
				} else{
					console.log("Index out of range!");
				}
				
			} else {
				console.log("Select a mailbox first.");
			}	
		} else if (line.toLowerCase().startsWith("q")){
			console.log("Goodbye!");
			await imap.disconnect();
			process.exit();
		} else{
			console.log("Command not known!")
		}
		process.stdout.write('imap> ');
	});
}


function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}] ${punctuation} `);
}

File: IPWorks Samples/IMAP Email Client/python/imap.py
# 
# IPWorks 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworks
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworks import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)



count = 0
mailboxes=[]

def fireMailBoxList(e):
  global count
  global mailboxes
  count+=1
  mailboxes.append(e.mailbox)

def fireMessageInfo(e):
  print(e.message_id + ") " + e.subject)

def fireTransfer(e):
  print(e.text)

def displayMailBoxList():
  global count
  global mailboxes
  for i in range(1,count+1):
    print(str(i)+") <%s>"%mailboxes[i-1])

def fireSSLServerAuthentication(e):
  e.accept = True

print("************************************************************************")
print("* This demo shows how to use the IMAP protocol to visit a mail server. *")
print("************************************************************************")

if len(sys.argv) != 4:
  print("usage: imap_email_client.py server username password\r\n")
  print("  server    the name or address of the mail server (IMAP server)")
  print("  username  the user name used to authenticate to the MailServer ")
  print("  password  the password used to authenticate to the MailServer ")
  print("\r\nExample: imap_email_client.py 127.0.0.1 username password")
else:
  try:
    imap = IMAP()

    imap.on_ssl_server_authentication = fireSSLServerAuthentication
    imap.set_mail_server(sys.argv[1])
    imap.set_user(sys.argv[2])
    imap.set_password(sys.argv[3])
    imap.on_mailbox_list = fireMailBoxList
    imap.on_message_info = fireMessageInfo
    imap.on_transfer = fireTransfer
    imap.connect()
    imap.set_mailbox("*")
    imap.list_mailboxes()

    command = input("Welcome! Type ? for a list of commands.> ")
  
    while True:
      if command == '?':
        print("IMAP Demo Commands")
        print("l                 list mailboxes")
        print("s mailbox_num     show messages in mailbox by index (e.g. 's 1')")
        print("r message_num     read the message by index (e.g. 'r 1')")
        print("q                 quit")
      
      elif str(command).startswith('l'):
        print("Mailboxes for " + sys.argv[2] + ":")
        count = 0
        mailboxes=[]
        imap.set_mailbox("*")
        imap.list_mailboxes()
        displayMailBoxList()
        print("Total: " + str(count))
      elif str(command).startswith('s'):
        index=str(command).split()[1]
        imap.set_mailbox(mailboxes[int(index)-1])
        imap.select_mailbox()
        messageCount=imap.get_message_count()
        if messageCount!=0:
          imap.set_message_set("1:" + str(messageCount))
          print("Messages in current mailbox:")
          imap.retrieve_message_info()
        else:
          print("No messages found in the current mailbox.")
      elif str(command).startswith('r'):
        print("Message text:")
        index=str(command).split()[1]
        imap.set_message_set(index)
        imap.retrieve_message_text()
      elif str(command).startswith("q"):
        imap.disconnect()
        print("Goodbye!")
        sys.exit(0)

      command = input("> ").lower()

  except IPWorksError as e:
    print("ERROR: %s"%e.message)



